source/common/http/conn_manager_impl.cc

```diff
@@ -199,11 +199,12 @@ void ConnectionManagerImpl::doEndStream(ActiveStream& stream) {
   // explicitly nulls out response_encoder to avoid the downstream being notified of the
   // Envoy-internal stream instance being ended.
   if (stream.response_encoder_ != nullptr &&
-      (!stream.state_.remote_complete_ || !stream.state_.codec_saw_local_complete_)) {
+      (!stream.filter_manager_.remoteComplete() || !stream.state_.codec_saw_local_complete_)) {
     // Indicate local is complete at this point so that if we reset during a continuation, we don't
     // raise further data or trailers.
     ENVOY_STREAM_LOG(debug, "doEndStream() resetting stream", stream);
-    stream.state_.local_complete_ = true;
+    // TODO(snowp): Removing this doesn't seem to break any tests? Can we remove this?
+    // stream.filter_manager_.state_.local_complete_ = true;
```

Comment:
@mattklein123 any idea if this would be safe to no longer set? We'd need to expose a setter for `local_complete_` on the FM in that case, which would be nice to avoid

Reply:
I honestly don't know and I would be very scared to change this code without fully understanding it. My advice would be to add a setter now with a clear TODO to figure out once the refactor is done if we need it.